packages/expo-document-picker/ios/DocumentPickerModule.swift

```diff
@@ -0,0 +1,203 @@
+    }
+
+    let pickingContext = PickingContext(promise: promise, options: options, documentPickerHandler: documentPickerDelegate)
+
+    picker.delegate = pickingContext.documentPickerHandler
+
+    // Store picking context as we're navigating to the different view controller (starting asynchronous flow)
+    self.currentPickingContext = pickingContext
+    currentViewController.present(picker, animated: true)
+  }
+
+  // MARK: - DocumentPickingResultHandler
+  func didCancelPicking() {
+    self.currentPickingContext?.promise.resolve(["type": "cancel"])
+  }
+
+  func didPickDocument(documentUrl: URL) {
+    guard let options = self.currentPickingContext?.options,
+    let promise = self.currentPickingContext?.promise else {
+      NSLog("Picking operation context has been lost.")
+      return
+    }
+
+    guard let fileSystem = self.appContext?.fileSystem else {
+      return promise.reject(Exceptions.FileSystemModuleNotFound())
+    }
+
+    // Cleanup the currently stored picking context
+    self.currentPickingContext = nil
+
+    do {
+      documentUrl.startAccessingSecurityScopedResource()
+
+      let fileSize = try getFileSize(url: documentUrl)
+      var newUrl = documentUrl
+
+      if options.copyToCacheDirectory == true {
+        let directory = URL(fileURLWithPath: fileSystem.cachesDirectory).appendingPathComponent("DocumentPicker")
+        let fileExtension = documentUrl.pathExtension
+        let path = fileSystem.generatePath(inDirectory: directory.absoluteString, withExtension: fileExtension.isEmpty ? fileExtension : ".\(fileExtension)")
+
+        newUrl = URL.init(fileURLWithPath: path)
+
+        do {
+          try FileManager.default.copyItem(atPath: documentUrl.path, toPath: newUrl.path)
+        } catch {
+          return promise.reject(UnableToCopyToCachesDirectoryException())
+        }
+      }
+
+      let filePathExtension = documentUrl.pathExtension
+
+      let mimeType = getMimeType(pathExtension: filePathExtension)
+
+      documentUrl.stopAccessingSecurityScopedResource()
+
+      return promise.resolve([
+        "type": "success",
+        "uri": newUrl.absoluteString,
+        "name": documentUrl.lastPathComponent,
+        "size": fileSize,
+        "mimeType": mimeType
+      ])
+    } catch {
+      return promise.reject(UnableToGetFileSizeException())
+    }
+  }
+}
+
+func getFileSize(url: URL) throws -> Int? {
+  let resources = try url.resourceValues(forKeys: [.fileSizeKey])
+  return resources.fileSize
+}
+
+func getMimeType(pathExtension: String) -> String {
+  if let uti = UTTypeCreatePreferredIdentifierForTag(kUTTagClassFilenameExtension, pathExtension as NSString, nil)?.takeRetainedValue() {
+    if let mimetype = UTTypeCopyPreferredTagWithClass(uti, kUTTagClassMIMEType)?.takeRetainedValue() {
+      return mimetype as String
+    }
+  }
+
+  return "application/octet-stream"
+}
+
+func mimeTypeToUTI(mimeType: String) -> String {
```

Comment:
Maybe enum can be used as mapping for mime-type <--> all kinds of UTI [like e.g. in image-picker]( https://github.com/expo/expo/blob/a0c5fbc58d259de9ccf742a217a0be266aeaa3ec/packages/expo-image-picker/ios/ImagePickerOptions.swift#L123) or [clipboard](https://github.com/expo/expo/blob/a0c5fbc58d259de9ccf742a217a0be266aeaa3ec/packages/expo-clipboard/ios/ClipboardOptions.swift#L13)

Reply:
Hey, I really like this approach, it looks really elegant and I think its fits nicely with `expo-image-picker` and `expo-clipboard`. However, I'm not sure if its the best approach for `expo-document-picker`. In `expo-image-picker` and `expo-clipboard` we have a very small subset of MIME types that are selectable and the enum mapping is perfect there - in `expo-document-picker` we also allow specifying `*/*`, but the provided MIME type can be something entirely different, e.g. `application/vnd.android.package-archive` for which the OS needs to initialize the `UTType`  separately. Doesn't that mean that we would also need to change the JS side to accept an enum instead of `string | string[]`? Maybe I'm missing something here 😅